### PR TITLE
Add new skill

### PR DESCRIPTION
**Pull request by GattieTime**
Added a button in profile creation to add a new skill to the database.

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -8012,6 +8012,11 @@
       "integrity": "sha512-iyam8fBuCUpWeKPGpaNMetEocMt364qkCsfL9JuhjXX6dRnguRVOfk2GZaDpPjcOKiiXCPINZC1GczQ7iTq3Zw==",
       "dev": true
     },
+    "ng-click-outside": {
```
Comment (IkenoXamos) [inline package-lock.json]: Was this dependency used in order to complete this feature?

File: src/app/Job-Service/job.service.spec.ts [hunk 2]
```diff
@@ -1,12 +1,60 @@
...
+        title: "Developer",
+        description: "test",
+        location: "Chicago",
+        isFilled: true,
+        skills: []
+        }];
+      let response;
+      spyOn(jobService, 'getAllJobs').and.returnValue(of(jobResponse));
+
+      jobService.getAllJobs().subscribe(res => {
```
Comment (IkenoXamos) [inline src/app/Job-Service/job.service.spec.ts]: Is this test actually doing anything? It looks like you're spying getAllJobs first, and telling it to automatically return jobResponse. And then you're expecting that the forced response matches jobResponse? This seems like you're effectively testing that your mock is returning what you told it to return.

File: src/app/Job-Service/job.service.spec.ts [hunk 3]
```diff
@@ -1,12 +1,60 @@
...
+        response = res;
+      });
+      expect(response).toEqual(jobResponse);
+    });
+
   });
 });
+
```
Comment (IkenoXamos) [inline src/app/Job-Service/job.service.spec.ts]: This test looks like it would also not actually be testing anything, if it weren't commented out. As is, why is this left in the code base as a comment? It should be either fixed or removed.

File: src/app/common/search-select/search-select.component.ts [hunk 4]
```diff
@@ -9,28 +9,60 @@ import { SkillService } from 'src/app/skill.service';
   styleUrls: ['./search-select.component.scss']
 })
 export class SearchSelectComponent implements OnInit {
-  constructor(private skillService: SkillService) {}
+  constructor(private skillService: SkillService,
+    private eRef: ElementRef) {
+    this.allSkills = [];
+    this.skillExistsEmitter.emit(this.skillExists);
```
Comment (IkenoXamos) [inline src/app/common/search-select/search-select.component.ts]: Why are you emitting skillExists in the constructor? I believe you only need to emit when the value has been modified, but in the constructor, nothing has happened yet that would have been able to modify the value.

File: src/app/common/search-select/search-select.component.ts [hunk 5]
```diff
@@ -9,28 +9,60 @@ import { SkillService } from 'src/app/skill.service';
...
+  skillExists = true;
+  @Output() skillTitleEmitter = new EventEmitter<any>();
+  searchTerm: string = "";
+
 
   selectOption() {
     this.emitSelections.emit(this.selections);
   }
 
+  // @HostListener('document:click', ['$event'])
```
Comment (IkenoXamos) [inline src/app/common/search-select/search-select.component.ts]: Why is this left as a comment? If it is not needed, it should be removed. Additionally, this appears to be the only use of HostListener. Since the only time it is used is commented out, then the import should also be removed.

File: src/app/common/search-select/search-select.component.ts [hunk 6]
```diff
@@ -9,28 +9,60 @@ import { SkillService } from 'src/app/skill.service';
...
+      this.skillTitleEmitter.emit(this.searchTerm);
+    } else {
+      this.skillExists = true;
+      this.skillExistsEmitter.emit(this.skillExists);
+    }
   }
 
   resetOptions() {
     this.options = this.allSkills;
+
```
Comment (IkenoXamos) [inline src/app/common/search-select/search-select.component.ts]: Blank line not really needed here.

Comment (IkenoXamos) [reply src/app/common/search-select/search-select.component.ts]: And the extra blank lines below aren't really needed, although it isn't that significant.

File: src/app/interview-service/interview.service.spec.ts [hunk 7]
```diff
@@ -1,12 +1,43 @@
...
+  interviewService =TestBed.get(InterviewService);
+});
 
   it('should be created', () => {
-    const service: InterviewService = TestBed.get(InterviewService);
-    expect(service).toBeTruthy();
+    // const service: InterviewService = TestBed.get(InterviewService);
+    expect(interviewService).toBeTruthy();
+  });
+  describe('retrieveInterviewById', () => {
```
Comment (IkenoXamos) [inline src/app/interview-service/interview.service.spec.ts]: This test doesn't do anything at all?

File: src/app/profile-create/profile-create.component.ts [hunk 8]
```diff
@@ -50,4 +52,29 @@ export class ProfileCreateComponent implements OnInit {
...
+  changeSkillExists(skillExists: boolean){
+    this.skillExists = skillExists;
+  }
+
+  onClickOutside(){
+    this.skillExists = true;
+  }
+
+  addSkill(){
+    console.log('Add Skill button clicked');
```
Comment (IkenoXamos) [inline src/app/profile-create/profile-create.component.ts]: Console.log not needed.

File: src/app/profile.service.spec.ts [hunk 9]
```diff
@@ -1,9 +1,11 @@
 import { TestBed } from '@angular/core/testing';
-
+import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';
```
Comment (IkenoXamos) [inline src/app/profile.service.spec.ts:2]: Is this import needed for this test file in it's current state?

File: src/app/skill.service.spec.ts [hunk 10]
```diff
@@ -1,12 +1,81 @@
...
+        {
+          id:2,
+          title:'CSS'
+        },
+        {
+          id: 3,
+          title: 'JavaScript'
+        }
+      ];
+      let response;
```
Comment (IkenoXamos) [inline src/app/skill.service.spec.ts]: Again, this series of statements looks like it is not actually testing anything.

File: src/app/skill.service.ts [hunk 11]
```diff
@@ -8,6 +8,9 @@ import { BehaviorSubject } from 'rxjs';
 })
 
 export class SkillService {
+  all() {
+    throw new Error("Method not implemented.");
```
Comment (IkenoXamos) [inline src/app/skill.service.ts:14]: I suppose this is fine for now, but this will of course need to be implemented once the backend is updated.

File: src/app/user.service.spec.ts [hunk 12]
```diff
@@ -1,12 +1,41 @@
...
+  describe('retrieveAllUsers', () =>{
+    it('should return all users ', () => {
+      const userResponse = [{
+        id: 0,
+        firstName: "Larry",
+        lastName: "King",
+        password: "",
+        username: "lking"
+      }];
+      let response;
```
Comment (IkenoXamos) [inline src/app/user.service.spec.ts]: Same concerns regarding tests. Is this actually doing anything?

**Review by IkenoXamos — CHANGES_REQUESTED** [review]
The actual functionality for adding a skill looks good, but the tests look like they are not effective.

**Review by IkenoXamos — COMMENTED** [review]
Making good progress, but more tests to fix.

File: src/app/profile.service.spec.ts (hunk 9 above)
Comment (GattieTime) [reply src/app/profile.service.spec.ts:2]: Yes, the default test fails without the dependency.

File: src/app/Job-Service/job.service.spec.ts [hunk 13]
```diff
@@ -1,12 +1,20 @@
...
-    const service: JobService = TestBed.get(JobService);
-    expect(service).toBeTruthy();
+  beforeEach(() =>{
+    TestBed.configureTestingModule({
+    imports: [HttpClientTestingModule],
+    providers: [JobService] 
   });
+  jobService = TestBed.get(JobService); //added for testing
 });
+
```
Comment (IkenoXamos) [inline src/app/Job-Service/job.service.spec.ts:15]: This can perhaps be in a different Pull Request, but this service should have more tests.

File: src/app/interview-service/interview.service.spec.ts [hunk 14]
```diff
@@ -1,12 +1,20 @@
 import { TestBed } from '@angular/core/testing';
-
+import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';
 import { InterviewService } from './interview.service';
 
 describe('InterviewService', () => {
-  beforeEach(() => TestBed.configureTestingModule({}));
+let interviewService: InterviewService;
```
Comment (IkenoXamos) [inline src/app/interview-service/interview.service.spec.ts:6]: This service should also have more tests in another PR.

File: src/app/shared/login.service.ts [hunk 15]
```diff
@@ -21,6 +23,27 @@ export class LoginService {
...
+
+  private extractData(res: any) {
+    let body = res.json();
+    return body || {};
+  }
+
+  private handleError(error: any) {
+    let errMsg: string;
+    try {
+      if(JSON.parse(error._body)) {
```
Comment (IkenoXamos) [inline src/app/shared/login.service.ts]: If statement not needed.

File: src/app/shared/login.service.ts [hunk 16]
```diff
@@ -21,6 +23,27 @@ export class LoginService {
...
+        console.log(error._body);
+        errMsg = JSON.parse(error._body);
+      }
+    } catch(e){
+      errMsg = 'There was an error logging in.';
+    }
+    if(errMsg === undefined || errMsg == ''){
```
Comment (IkenoXamos) [inline src/app/shared/login.service.ts]: errMsg shouldn't able to be undefined once the above if statement is removed.

File: src/app/Job-Service/job.service.spec.ts (hunk 13 above)
Comment (GattieTime) [reply src/app/Job-Service/job.service.spec.ts:15]: We have set up a new branch for tests outside the scope of the components/services we directly touched with the add skill functionality. We will make a separate pull request when those tests are ready.

File: src/app/interview-service/interview.service.spec.ts (hunk 14 above)
Comment (GattieTime) [reply src/app/interview-service/interview.service.spec.ts:6]: We are working on those tests on another branch. See above.

**Review by IkenoXamos — APPROVED** [review]
Looks good.
